src/App.tsx

```diff
@@ -16,35 +17,98 @@ export const goodsFromServer = [
 ];
 
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState<SortType | ''>('');
+  const [isReversed, setIsReversed] = useState<boolean>(false);
+
+  enum SortType {
+    ALPHABET = 'alphabet',
+    LENGTH = 'length',
+    RESET = 'reset',
+  }
+
+  const visibleGoods = [...goodsFromServer].sort(
+    (good1: string, good2: string) => {
+      switch (sortField) {
+        case SortType.ALPHABET:
+          return good1.localeCompare(good2);
+        case SortType.LENGTH:
+          return good1.length - good2.length;
+        case SortType.RESET:
+          return 0;
+
+        default:
+          return 0;
+      }
+    },
+  );
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  const reverse = () => {
+    setIsReversed(prevState => !prevState);
+  };
+
+  const reset = () => {
+    setSortField('');
+    setIsReversed(false);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={classNames('button is-info', {
+            'is-light': sortField !== SortType.ALPHABET,
+          })}
+          onClick={() => {
+            setSortField(SortType.ALPHABET);
+          }}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={classNames('button is-success', {
+            'is-light': sortField !== SortType.LENGTH,
+          })}
+          onClick={() => {
+            setSortField(SortType.LENGTH);
+          }}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={classNames('button', 'is-warning', {
+            'is-light': !isReversed,
+          })}
+          onClick={reverse}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortField || isReversed) && (
+          <button
+            type="button"
+            className={classNames('button', 'is-danger', 'is-light')}
+            onClick={reset}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods.map(good => (
+          <li key={good} data-cy="Good">
+            {good}
+          </li>
```

Comment:
Using the good's name as a key is not reliable if there can be duplicates. It's better to use a unique identifier like an id. If the goods' names are guaranteed to be unique, this can be acceptable, but it's still a good practice to use an id whenever possible. [CODE KNOWLEDGE]